tools/apiview/parsers/cpp-api-parser/ApiViewProcessor/ProcessorImpl.cpp

```diff
@@ -0,0 +1,520 @@
+    {
+      subdirectories.push_back(entry.path());
+      auto inner = GatherSubdirectories(entry.path());
+      subdirectories.insert(subdirectories.end(), inner.begin(), inner.end());
+    }
+  }
+  return subdirectories;
+}
+std::string replaceAll(
+    std::string_view const& source,
+    std::string_view const& oldValue,
+    std::string_view const& newValue)
+{
+  std::string newString;
+  newString.reserve(source.size());
+  size_t findPos{};
+  size_t lastPos{};
+
+  while (std::string::npos != (findPos = source.find(oldValue, lastPos)))
+  {
+    newString.append(source, lastPos, findPos - lastPos);
+    newString += newValue;
+    lastPos = findPos + oldValue.length();
+  }
+  newString += source.substr(lastPos);
+  return newString;
+}
+
+const nlohmann::json JsonFromConfigurationPath(
+    std::string_view configurationRoot,
+    std::string_view const& configurationFileName)
+{
+  std::filesystem::path configurationFilePath{configurationRoot};
+  configurationFilePath /= configurationFileName;
+  std::ifstream configurationFile{configurationFilePath};
+  if (!configurationFile.is_open())
+  {
+    throw std::runtime_error(
+        "Unable to open configuration file: " + configurationFilePath.string());
+  }
+  nlohmann::json configurationJson;
+  configurationFile >> configurationJson;
+  return configurationJson;
+}
+
+ApiViewProcessorImpl::ApiViewProcessorImpl(
+    std::string_view directoryToProcess,
+    std::string_view const& configurationFileName)
+    : ApiViewProcessorImpl(
+        directoryToProcess,
+        JsonFromConfigurationPath(directoryToProcess, configurationFileName))
+{
+}
+
+ApiViewProcessorImpl::ApiViewProcessorImpl(
+    std::string_view directoryToProcess,
+    nlohmann::json const& configurationJson)
+    : m_currentSourceRoot{std::filesystem::absolute(directoryToProcess)},
+      m_classDatabase{std::make_unique<AzureClassesDatabase>(this)}
+{
+  // CHDIR to the directory to process so relative paths in the configuration are properly resolved.
+  auto currentDirectory{std::filesystem::current_path()};
+  std::filesystem::current_path(directoryToProcess);
+  if (configurationJson.contains("includeInternal"))
+  {
+    m_includeInternal = configurationJson["includeInternal"];
+  }
+  if (configurationJson.contains("includeDetail"))
+  {
+    m_includeDetail = configurationJson["includeDetail"];
+  }
+  if (configurationJson.contains("includePrivate"))
+  {
+    m_includePrivate = configurationJson["includePrivate"];
+  }
+  if (configurationJson.contains("filterNamespace")
+      && !configurationJson["filterNamespace"].is_null())
+  {
+    if (configurationJson["filterNamespace"].is_string())
+    {
+      m_filterNamespaces.push_back(configurationJson["filterNamespace"].get<std::string>());
+    }
+    else if (configurationJson["filterNamespace"].is_array())
+    {
+      m_filterNamespaces = configurationJson["filterNamespace"];
+    }
+    else
+    {
+      throw std::runtime_error(
```

Comment:
For reference to above comment, if we throw from here the cwd doesn't get reset.